### ISSUE TITLE
Dispersy mini: Dispersy no longer stores any data in databases

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#2778]
Remove the _double storage principle_ of Tribler. Dispersy becomes a state-less and clean overlay.

Each Dispersy community is now responsible for their own storage. We now have significant overlap between the tribler and dispersy database. Move much of this responsibility in several iterations upwards. Downwards we move the low-level connectivity stuff. The walker becomes responsible for storing all public keys, interaction records, IP addresses, etc.
See: https://github.com/Tribler/dispersy/issues/61

![libtribler_architecture](https://cloud.githubusercontent.com/assets/325224/22511176/14168b2a-e895-11e6-9aa1-22274403d08c.png)

We have several issues to reach ```dispersy mini``` and refactor 14000 Dispersy lines. Braindump :
  - [ ] community.py document current API and future community API
  - [ ] requestcache.py replace fake sessions with real sessions or remove
  - [ ] candidate.py replace with a proper walker, integrate MultiChain to provide trusted peer discovery
  - [ ] crypto.py and authentication.py replace with RFC-based DTLS and DH hadshakes
  - [ ] distribution.py remove non-scalable Full sync, pruning, and LastSyncDistribution and replace with primitives based on transitive trust and locality with unbounded scalability. ToDo: specify N-Hop-Multichain-Sync
  - [ ] conversion.py replace with protocol buffers
  - [ ] dispersy.py refactor out of existance: should not deal with details of IPv4 address WAN voting, communities and duplicate sync messages
  - [ ] dispersydatabase.py kill with fire and refactor the double storage principle (very hard, see prior abandon PRs, Axiom https://github.com/Tribler/tribler/compare/devel...LipuFei:refactory_axiom_db?expand=1, Storm https://github.com/Tribler/tribler/pull/2416)

```
wc -l Tribler/dispersy/*.py
   397 Tribler/dispersy/authentication.py
   298 Tribler/dispersy/bloomfilter.py
   335 Tribler/dispersy/candidate.py
  3703 Tribler/dispersy/community.py
  1296 Tribler/dispersy/conversion.py
   424 Tribler/dispersy/crypto.py
   417 Tribler/dispersy/database.py
    99 Tribler/dispersy/destination.py
   705 Tribler/dispersy/dispersydatabase.py
  2273 Tribler/dispersy/dispersy.py
   325 Tribler/dispersy/distribution.py
   376 Tribler/dispersy/endpoint.py
    17 Tribler/dispersy/exception.py
     0 Tribler/dispersy/__init__.py
   219 Tribler/dispersy/member.py
   615 Tribler/dispersy/message.py
    46 Tribler/dispersy/meta.py
   671 Tribler/dispersy/payload.py
   234 Tribler/dispersy/requestcache.py
   112 Tribler/dispersy/resolution.py
   393 Tribler/dispersy/statistics.py
   124 Tribler/dispersy/taskmanager.py
   458 Tribler/dispersy/timeline.py
   335 Tribler/dispersy/util.py
 13872 total
```

Roadmap note: instead of tweaking this layer for years and having fun, we need to be pragmatic. So no 10 month refactoring job and 7 months of nasty bug hunting. Clean-up and move on to real productive work within 6 months!

Good luck!

**Comment by egbertbouman** [comment]
Maybe would even look at the byte stats that go through the socket (with tunnel prefix?) instead of using notifications from the TunnelCommunity. Just an idea.

**Comment by qstokkink** [comment]
If we want to be pragmatic and fast, we should follow a concrete plan. So I hereby present:

<hr>

### Quinten's 6 month concrete action plan

I believe the best way to approach this, is starting by refactoring out the current Dispersy walker into a separate module. Then we would write the new walker interface as we would like it to be in the future. If this does not fit the old interface, the old walker would have to be adjusted to keep everything functional.

This is the first milestone which can be merged into the repository.

At this point the new walker can be created. It would consist (roughly) of the following 3 components, which can be implemented (more-or-less) concurrently:

1. A peer management system (public key to IP address, reputation, etc. - covers candidate.py, member.py, database.py)
2. A peer discovery system (puncture, authentication, id etc. and conversion.py)
3. A high-level trust based mechanism implementation for delivering peers through the interface (multichain community)

Once this is done, another milestone would be able to be merged into Dispersy: dropping the old walker and using the new one. At this point backward compatibility will probably be broken (if the peer discovery system uses Protocol Buffers messages already).

After this, we should probably finish off with refactoring the requestcache.py, community.py and dispersy.py.
<hr>

Feedback or alternatives are welcomed.

**Comment by devos50** [comment]
Just a mental note here: I think that Dispersy communities should adhere to the separation of concerns principle where each community has it's own functionality and responsibilities. The communication between these communities should be as little as possible i.e. they should be able to run in isolation.

**Comment by synctext** [comment]
Dispersy-identity message: remove and always include full public key.

**Comment by egbertbouman** [comment]
@devos50 I agree, but as far as I know communities already run separately. Are you referring to some specific instance where communities communicate with each other where they shouldn't?

@synctext I completely agree with removing the dispersy-identity/dispersy-missing-identity message stuff. However, note that this requires upgrading all communities to v2 (allchannel,channel,multichain! are using v1), so the old communities would no longer be accessible. If we are going to do this, we may as well use curve25519/ed25519 members everywhere. Then we can remove the M2Crypto dependency. Also a good time to switch to protocol buffers...

**Comment by synctext** [comment]
@egbertbouman  Yes to V2, yes for curve25519, yes to protocol buffers. Cleanup year..

**Comment by qstokkink** [comment]
The first refactoring is in: https://github.com/qstokkink/py-ipv8.

**Comment by qstokkink** [comment]
Small update: just now the first bytes were sent to - and received from - a tracker, marking the first backwards compatible communication.

Notable changes between IPv8 and Dispersy so far are (1) the removal of the `M2Crypto` dependency, (2) the removal of `authentication.py`, `conversion.py`, `destination.py`, `distribution.py` and `resolution.py` and (3) removal of meta objects and dramatic downsizing of `message.py`.

We still have quite a way to go (many messages to convert to the new system) for full on backwards compatibilty though.

**Comment by qstokkink** [comment]
Now that I can communicate with Dispersy, I had some fun and captured a little under 10 seconds of peer discovery:

![output](https://user-images.githubusercontent.com/3630389/27956213-0ae41fbe-6319-11e7-803b-b7172935b042.png)

**Comment by qstokkink** [comment]
Some internal imagery for my own convenience:
![20170707_172047](https://user-images.githubusercontent.com/3630389/27964557-fbaeb2c8-6338-11e7-85d5-4b832c02cdfe.jpg)
<sub>
Green: hole punching package
Blue: peer discovery package
Red: pitfall
Black: note
</sub>

**Comment by qstokkink** [comment]
Cool update: the basic (no hidden services) TunnelCommunity is integrated and working in IPv8.

**Comment by qstokkink** [comment]
The first test is up and running for the TunnelCommunity 🎉 . Performance is notably better.

Doing the following:
- Create two nodes with their own keypairs
- Have each of them start a TunnelCommunity
- Exchange an introduction request and response (+ assert)
- Create a circuit of length 1 (+ assert)
- Unload the two communities

Now takes 0.03~0.04 seconds, instead of 20 seconds.

**Comment by synctext** [comment]
making impressive progress.. 40msec, interesting speedup metric.

Obviously I'm anxiously awaiting if IPv8 also has an impact on Bittorrent tunnel MBps.. But cleaning, refactoring, and bugfree code is what Tribler needs first.

**Comment by qstokkink** [comment]
@synctext I think MBps will go up a little bit due to the cleaner back-end. It won't be as big of a boost as using multiple sockets/multiprocessing though.

Also, progress:
- The test suite has gotten pretty mature now (lots of lock-up detection and stuff, but still fast).
- I got the hidden services working consistently instead of 50/50 now (the community had a particularly nasty race condition in it, related to the KeyRelayCache and KeyRequestCache request caches sharing the same name).
- Now I will move on to testing and validating the TrustChain community.

**Comment by synctext** [comment]
year 2018 toughts.. move whole Libtorrent API into a storage wrapper. becomes part of the IPv8 stack API: trust,identity, secure communication and swarm-based storage.
Merely 1 bulk and 1 control socket for all DHT, miniDispersy, swarming and crypto matters. not a priority, but clean vision.

**Comment by qstokkink** [comment]
More progress: TrustChain now also has tests and appears to be working correctly. Now I'll move on to the boring stuff like (1) hooking up a coverage check to see if I am missing lines in my tests, (2) writing a README.md with all the depedencies, (3) writing docstrings and other assorted documentation and (4) run the tests on other platforms.

Overall, IPv8 has all of the **beta 1 functionality complete**:
 - *Public service discovery* (formerly Dispersy DiscoveryCommunity)
 - *Hidden service discovery* (formerly HiddenTunnelCommunity)
 - *Verified and unverified messaging* (formerly Dispersy)
 - *Anonymous messaging* (formerly TunnelCommunity)
 - *Simple TrustChain* (formerly TrustChainCommunity)
 - *Peer discovery*: random and edge walks (formerly Dispersy and TrustChainCommunity)
 - *Peer keep-alive* (formerly spread over multiple communities)

IPv8 does not (yet) include:
 - *Mixnet functionality*
 - *Hidden attribute attestation*
 - *Extended TrustChain*
 - *Stream based messaging* (instead of packet based)

At some point, all of this functionality should probably be shaped into some nice API.

**Comment by synctext** [comment]
congrats! now the boring part..

Do we need unverified messages in the future? (no idea what they really are..) crypto==NULL and no authentication could be avoided.

**Comment by qstokkink** [comment]
@synctext What I mean with *verified* is that the messages are correctly signed (message public key and message signature match). In the past Dispersy also added replay attack prevention, but this ate disk space and CPU time and wasn't used in the tunnel and TrustChain communities. So I left it out in IPv8 (for now).

The TunnelCommunity and the TrustChainCommunity use their own verification mechanism. So, for the more advanced communities unverified messages are also available. Furthermore, the messages are smaller and faster to construct, making them convenient for NAT hole punching.

I do agree that this is a potential security risk for first-time community developers. That is why I made verified messages the default option.

**Comment by qstokkink** [comment]
Testing update: now on 82% **branch** coverage, 132 tests, time in tests 5.721 seconds, time with overhead (test suite/script setup and teardown) 14.496 seconds. Moving on to other boring parts.

**Comment by qstokkink** [comment]
We are now sporting a README: https://github.com/qstokkink/py-ipv8

**Comment by qstokkink** [comment]
Ok, I also added missing documentation, which completes the feature set for *beta 1*. Anyone should now be able to make full use of IPv8's functionality. Time for 4 days of vacation 🌦 😎.

**Comment by synctext** [comment]
As IPv8 is including web-of-trust and authentication features, please read: https://doi.org/10.6028/NIST.SP.800-63-3
This documents the hard-core NIST guidelines on Digital Identity:
```
IAL2: Evidence supports the real-world existence of the claimed identity and verifies that the
applicant is appropriately associated with this real-world identity. IAL2 introduces the need for
either remote or physically-present identity proofing. Attributes can be asserted by CSPs to RPs
in support of pseudonymous identity with verified attributes.
```

**Comment by qstokkink** [comment]
For my own convenience, this is the executive summary of the NIST identity standards:

| IAL | Description |
| --- | --- |
| 1 | Self-assert |
| 2 | IAL1 + third party assert + pseudonyms |
| 3 | IAL2 + third party attribute attestation + physical presence |

| AAL | Description |
| --- | --- |
| 1 | Some authentication protocol |
| 2 | Multifactor (at least 2) authentication |
| 3 | AAL2 + hardware authenticator + authenticated verifiers |

| FAL | Description |
| --- | --- |
| 1 | Signature from federation |
| 2 | Encrypted signature from federation for RP |
| 3 | FAL2 + AAL1<sup>+</sup> |

**Comment by synctext** [comment]
Neat trick to possibly implement for 2018: as measure of last bootstrap failure resort...
join the hard-coded #IPv8-bootstrap channel and meet others just like you and talk to them.
Inspired by the 8 Bitcoin bootstrap features: https://en.bitcoin.it/wiki/Satoshi_Client_Node_Discovery

**Comment by synctext** [comment]
[Supersingular Elliptic Curve Isogenies](https://github.com/defeo/ss-isogeny-software) is a quantum-resistant cryptosystem. How mature is this work and their Python? Realistic for a 2018 roadmap issue?

**Comment by qstokkink** [comment]
Alright, I fed the same Community to both Dispersy and IPv8 and had two nodes exchange 10 ping-pong messages with each other. IPv8 is not even in the same ballpark as Dispersy:
![output](https://user-images.githubusercontent.com/3630389/29996238-5208edb6-8ffb-11e7-8ed3-c54918a20be1.png)
Do bear in mind that these are the fastest message settings available for both platforms. Also note that Dispersy was designed for bandwidth efficiency and IPv8 for throughput.

**Comment by synctext** [comment]
@qstokkink nice!
Upcoming Tuesday we hope to be able to present a "messages per second" graph of IPv8 for various settings or just a single number. Plus a number for Dispersy. Performance regression testing would be great to have from this early stage onwards [see this amazing example of professional performance testing](http://speed.pypy.org/).

**Comment by synctext** [comment]
btw if a master student showed me your 10 ping-pong speed graph I would send them away to redo it :-)

The linear part from message 3 until message 10 from Dispersy node 1 and 2 is **perfect** linear, although these are really 7 unique incoming message events taking 15 seconds.

**Comment by qstokkink** [comment]
@synctext I was also expecting more of a "usual" cumulative distribution function with buffers filling up etc. The main cause for it being so linear is probably the fact that the ping<->pong is synchronous (each ping<->pong waits for the last to finish).

Also, there is probably some hidden 'feature' in Dispersy causing a forced 1 second delay between transmitting two of the same message (it's not actually perfectly linear there is a variance of a few milliseconds).
You can compensate for this 'feature' by taking the time difference between a pong and the next synchronous ping of the couterparty instead. This puts Dispersy at a time of roughly 21ms/message and IPv8 at roughly 2ms/message. 

Given these results I would also say that (1) it indeed makes more sense to just present throughput as a number and (2) we should just test sending messages asynchronously to give Dispersy a fighting chance.

**Comment by qstokkink** [comment]
Spurred on by the above results I decided to rip out all of the endpoint nonsense from Dispersy and simply exchange messages between communities directly. This is now an unfair comparison as IPv8 is actually sending through its endpoints, but it is still faster:
![output](https://user-images.githubusercontent.com/3630389/30001964-5c930d0e-909e-11e7-8181-ab87228a72c6.png)
Note that this is still synchronous communication. I named this supercharged Dispersy version `Dispersy+` in the graph.

**Comment by qstokkink** [comment]
Alright here is some publishable output of large experiments:
![manualep](https://user-images.githubusercontent.com/3630389/30053335-b0990d50-9228-11e7-8c0e-04849f3b96b4.png)
![standaloneep](https://user-images.githubusercontent.com/3630389/30053336-b2a3df08-9228-11e7-9576-743eab8502df.png)
These are box-and-whisker plots of the round-trip time of messages in Dispersy vs IPv8 using (1) no intermediate routers and (2) 1 intermediate router. Like before, this is a polished up/faster version of Dispersy.

**Comment by synctext** [comment]
[IPv10 trolling at IETF](https://news.ycombinator.com/item?id=15190097) but has good intro [text](https://tools.ietf.org/html/draft-omar-ipv10-06):
```
- Three short-term solutions (CIDR, Private addressing, and NAT) were
  introduced in the mid-1990s but even with using these solutions,
  the IPv4 address space ran out in February, 2011 as announced by
  IANA, The announcement of depletion of the IPv4 address space by
  the RIRs is as follows:

  * April, 2011:      APNIC announcement.
  * September, 2012:  RIPE NCC announcement.
  * June, 2014:       LACNIC announcement.
  * September, 2015:  ARIN announcement.

- 19 years have passed since IPv6 was developed, but no full migration
  happened till now and this would cause the Internet to be divided
  into two parts, as IPv4 still dominating on the Internet traffic
  (85% as measured by Google in April, 2017) and new Internet hosts
  will be assigned IPv6-only addresses and be able to communicate with
  15% only of the Internet services and apps.
```

**Comment by qstokkink** [comment]
@synctext yes, this is hilarious: https://twitter.com/The_Road_Series/status/868140346177355777

**Comment by qstokkink** [comment]
On the IPv8 attestation component:
As it turns out, calculating the Weil pairing on a finite field Elliptic curve using a distortion field is kind of difficult. Only one real implementation is known to exist and it is written in Mathematica. So, to facilitate the Python implementation, I got myself a Mathematica license from the TU Delft and formalized the Mathematica implementation as pseudocode:
[formulae.pdf](https://github.com/Tribler/tribler/files/1300026/formulae.pdf)
Other than this, the rest of the implementation should be smooth sailing.

**Comment by synctext** [comment]
impressive! A 4 page algorithm. Guess we really want to line up a big pile of testcases to compare both implementations eventually.

**Comment by qstokkink** [comment]
Making progress translating the Mathematica code into Python. 
![afbeelding](https://user-images.githubusercontent.com/3630389/30473849-03225ba4-9a02-11e7-8071-2ff5fb22519f.png)

Implementations are rare, no open source library is feature complete.
The only other library available has a performance of:

```
The PBC library is designed to be the backbone of implementations of pairing-based cryptosystems,
thus speed and portability are important goals. It provides routines such as elliptic curve
generation, elliptic curve arithmetic and pairing computation. Thanks to the GMP library, despite
being written in C, pairings times are reasonable. On a 1GHz Pentium III:
Fastest pairing: 11ms
Short pairing: 31ms
```

[Note that Type B is not implemented in this library](https://crypto.stanford.edu/pbc/manual/ch08s04.html). However, this is critical for the implementation of the Weil pairing (Abelian mapping) for finite field Elliptic curves using distortion maps.

**Comment by qstokkink** [comment]
I needed coefficient based operators for the Python implementation, which I worked out here (mostly for my own reference):
![ffformulea](https://user-images.githubusercontent.com/3630389/30588314-8338cd96-9d36-11e7-80f8-318391931838.png)

**Comment by devos50** [comment]
Idea (came up during meeting): create an identity validation block in your TrustChain.

**Comment by qstokkink** [comment]
Alright, the Weil pairing seems to be functional now (Python produces the same value as Mathematica):
![preview](https://user-images.githubusercontent.com/3630389/30699335-c0593d96-9ee4-11e7-8a2a-bdd7b376f3c0.png)
(Mathematica is above the black bar, Python is under the black bar)

**Comment by qstokkink** [comment]
Attestation update, I can now perform the Stanford encryption. Here is some raw output:

```
CIPHERTEXTS (ENCODED BIT VALUE: 1):
4467 + 75x
3019 + 4629x
2318 + 4070x
2318 + 4070x
1088 + 1834x
1318 + 2225x
4203 + 1184x
3534 + 2558x
3019 + 4629x
2318 + 4070x
1318 + 2225x
4061 + 1743x
4392 + 5738x
1088 + 1834x
976 + 3255x
976 + 3255x
5067 + 3979x
5067 + 3979x
63 + 5237x
4467 + 75x

DECODED (CIPHERTEXT EXPONENTIATION WITH SK):
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
```

**Comment by synctext** [comment]
Great doc examples:
- Related work from IPFS, we call it PVv8, they call it libp2p : https://github.com/libp2p/specs/blob/master/4-architecture.md#4-architecture
- Great looking roadmap, without any real details https://maidsafe.net/#roadmap
- boring storage coin roadmaps: http://blog.storj.io/post/159566947133/storj-new-protocol-new-opportunities
- less is more: http://sia.tech/about/milestones/

**Comment by qstokkink** [comment]
After some major optimizations: managed the first 512-bit-primes-key generation. Still some room left for speedup. Encryption/decryption itself is pretty fast.

```
RANDOM CIPHERTEXTS:
27819787572934406955043645093393842401993071469366681292238424796137619677306643704822401176837845947645204143850293453891591259579432512637721892788528947 + 127269305053866411856240263656167770550683818740502830977586904429068538960419847382428445626306588169955492852281714997188187203300826967806657664100386085x
121320047998478831289340101355181059721687546361944999842911173946274855846740576508835311135677350097401565566500597307401060517315873749593281997325211185 + 992665056501391192073459270486308812253456417209749851122631094903993455368064223832184992685028381073760637676338743854592830166268112690103770515787088x
165726942691611537722565503796304737203706713276071135536436385769342229544214397727932176683554222107632984298683789248179118639577498865888136382261567108 + 157981475677911198276430776020436291396737662344945486541624376254311901520421268911513231119734547280233950160326096430574824182053524070957352877914390203x
130200020437030443544603127176135889224888771494049811957338425255413285082108104855732607682362196233329342927242209686602158047616746305848018292520711198 + 132226935570814425704817021574265878850011584614806793295121561638924729783155632379015289305596310697652941681758402857655796258409317249088508406774493176x
110719948931283941057187983691216140403502429301765269439516963474941613604084883765420684437779497587737016002276630820468184617712931968441709253554338703 + 60370133285663657060401187796729473649240385523812726231277187547826580731991976511524927508285634624107970657913672507046750069321843085770165183862120155x
60652512444857048143098802167520153984106429919197478769828080742343336036642581067893127045534660471801170453490067391764835883979185761152398633272722296 + 46905838323311527982788437853097496490405274315352299117203260928536230764240664226596160478447508959549929193823392085873516411326405319290104775956058288x
66165640286271503179560314555603371912689710010813730692678814176529489335148211794322595647537743595087106348292111183020170970602269205326336585136077977 + 84845234921119080616624114880680483221074918260088667086855131218672763883656947087990370652102649458956895152076020617730362634954839970824753571025851212x
110682688728521575412758157767852881565718286772888923800500319599995110639261152305448742802048614322157332759789589414643355903097395498285532000163152076 + 119615694458238726986906378850051493185090803986815471090937147976779480888462315913299990289707636915395192905101200790759249175203345556644941994139851335x
165970072031676447730638238197191305516211228227854547150412912143639617564931467583799628730062799195340040423629935744750773992167459601608267352620111173 + 103214293313527887854043870308368258982524781740017606269573765609598664885292255104086760925690704177075858611417274898765475646618240491777422402349883664x
77497255732997188725991095890352485981599596699104475860171469784130169470584820932129009210911014419713539646544516310742667472897876341854503019039145005 + 73514075927872420531775351073912512956003967378501056795679890191010660234819259991010683932051230528639714962547149365034975941038126831701971335118475324x
59098145107605158921046050296354854412778472144398939942945658504189233290320403048726759691714746395864572084219318631161238505105760344010024244697020284 + 94929046558930929094187164600200273555499860427264253446115533927906013582742021193406356557070146550343821486284117887655821795569539853342891138760948831x
28414691139304030668669687052297775523533775487133723385101025862281651075666806795190038766397114259509309098811139037205869907680396160493104331089996949 + 106220645424240243588172370987508966386281324914639081960321926504970664098696062081166187313427343050793121466531534230771199145391243751570754963273429321x
123774743349305464935212416574016236445211208767843718367576419920059679632051231647078082860690936289877238858080827262694036400770440377500783533273787883 + 35571857516981593067757520372604831936845092063992518466151119674769687775690683526232862435528988650127411223593213727625737704484339099103213094456874418x
42944656913437718781153554975259492048986385633566985221615601337209481418141625087417673148108659075013760373975618763505078978766146778905747320510511522 + 40499550451974623016246374636960043598776271751443633630666531273847067370258777945089279550996120615107500493453265835981931698784463119417792452226667955x
106957416702359616794844470395410297067474379336183917441899059577039586635264700166259201350460898114443268581574321495977478664919064518686504560370492225 + 71897464142060450030931507674757648224533678212844544249277635207020356025278141154625175407148177163768807294250392284497092951194447997506132149678552714x
123275424588819998626435222785112162149769503581580007550650434387525567165495198037168879318101936672165633149135505165757847799832270858543973361820357245 + 59117066022973382699618728563111733089869773923249293176984415581027564901768740504406543631210788811136340586261158274694300791805522738795895971688280238x
63341630495795797332212197850122580710036950544082647555788844298412831996024231488964687061168546157283373961711964636045510876004675343449972440230316465 + 112585731153913932063162836579066385241845518424913751013043871121554112940229095197283536850495465591797614509743983956393383051685763059733183376048815124x
186728210054630030123954257128454460219908580197330657739943249315835989785844583725141425562044718725203936882712883949340142113334964166945341064459558271 + 60876517965846457983179101940302301528889925872532375001149237067283904574021039465900094936284484478229468766752470659281549946844504373490724208380403597x
99140531571271884859545654324786630837061714201360583296118192083407488219291627113492161824026747782899343818171287711090639904463459896102360641108213613 + 167327457578734220805264616363014055223891136745094941713074710793062374502861320792064736650852539205189147429862759906956667139329487504809678079664398255x
69718824430869421624048666651720723923912832179594072390321198837663068249741181123466104223262477188461231690586597100634019567588013134339643128409113556 + 59131644368552319847780699908874772101063559350860391205989075222394093198082077193243574986702587300347155499680747729971154973271917641303541701449222000x
----------------------------------------
SHOULD ALL BE 1:
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
```

**Comment by synctext** [comment]
Please read IPv8 related work from IETF: https://tools.ietf.org/html/draft-padma-ideas-problem-statement-03
different storyline, but in-scope for IPv8.

**Comment by qstokkink** [comment]
Status update: I optimized the code enough that I can generate a key using 1024 bit primes in about 20 seconds. It's still not fast, but its secure and usable. I'll move on to implementing the actual zero-out scheme on top of the cryptosystem now.

**Comment by qstokkink** [comment]
I have generic attribute attestation now, here is the output of a full proof of a 64 bit number:
```
Expected:	{0: 20, 1: 9, 2: 3, 3: 0}
Got:		{0: 20, 1: 9, 2: 3, 3: 0} after 32 challenges
```
I also want to add support for proving ranges (which should be somewhat easy). After that I'm going to do a big code cleanup and then merge it into IPv8.

**Comment by synctext** [comment]
Excellent progress..

btw fast prime generation bug
http://www.reuters.com/article/us-infineon-cyber/infineon-says-has-fixed-encryption-flaw-found-by-researchers-idUSKBN1CL2KC?feedType=RSS&feedName=technologyNews

**Comment by qstokkink** [comment]
Slight change of plans: I did the code cleanup today and I'll integrate this into IPv8 next. I'll leave the range proofs for after the MVP has been released.

**Comment by synctext** [comment]
Cardinal requirement for Dispersy re-write: __Both metadata and content will move as fast as the underlying network permits__ (see example requirement list https://github.com/ipfs/faq/issues/47)

**Comment by qstokkink** [comment]
Small milestone: code moved from [the secret project repository](https://github.com/qstokkink/secret-project) to [the IPv8 repository](https://github.com/qstokkink/py-ipv8/tree/master/ipv8/attestation/wallet).

**Comment by qstokkink** [comment]
Larger milestone: unit tests are done for the basic attestation community, now I just need a way to advertise attestation hashes and metadata: I was thinking of using TrustChain for this.

**Comment by synctext** [comment]
Trustchain for transparency and tamper-proofness.
Certificate transparency is "a big thing" with [epic past failures](https://www.wired.com/2011/09/diginotar-bankruptcy/). In 2013 [Google wrote RFC6962](https://tools.ietf.org/html/rfc6962) about it.

**Comment by qstokkink** [comment]
Ok, IPv8 now has all the basic functionality for creating, sharing and verifying generic attestations.

I'll switch to creating documentation and turning IPv8 into a standalone service now.

**Comment by qstokkink** [comment]
I realized today that I could make the Boneh scheme more secure by not having it depend upon the factorization problem. Long story short; by simply leaving out `n` from the public key the system becomes 2^512 times stronger.

**Comment by qstokkink** [comment]
IPv8 can now be started and stopped as a service, as per request from @devos50 .

**Comment by qstokkink** [comment]
I created an artistic impression of the current stack, in table form:
![ipv8 id](https://user-images.githubusercontent.com/3630389/32173534-a2ea4818-bd7f-11e7-89e2-93d358939635.png)

**Comment by qstokkink** [comment]
(Even though the backend is fully functional) the IPv8 app/GUI is still a W.I.P., here is the design:
![img_20171215_113616](https://user-images.githubusercontent.com/3630389/34038674-7254d51a-e18d-11e7-8e96-f3f53a3514bd.jpg)

**Comment by qstokkink** [comment]
IPv8 now has a REST API:
![afbeelding](https://user-images.githubusercontent.com/3630389/34104394-88ed2fba-e3f0-11e7-81e7-927bce916694.png)

**Comment by qstokkink** [comment]
Getting there with the GUI:
![screenshot_20171220-112554](https://user-images.githubusercontent.com/3630389/34202704-e272b968-e578-11e7-88ab-29abbbd14a20.png)

**Comment by qstokkink** [comment]
Image dump of the working Android app with a debug data:

<table>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263041-7c7524a6-e66e-11e7-9230-4c6ff9571545.png">
</td>
<td>
</td>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263042-7e2ea89e-e66e-11e7-816c-b818036c8695.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263051-801f0388-e66e-11e7-9df5-e86099e268e1.png">
</td>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263053-83384e8a-e66e-11e7-909e-94dc9fc9c4bd.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263056-852d800c-e66e-11e7-82e4-647f8227efd7.png">
</td>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263059-8741db22-e66e-11e7-93f7-fa0e7a234ee8.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/3630389/34263063-885a774e-e66e-11e7-9f36-e9e7a2f4ce57.png">
</td>
</tr>
</table>

**Comment by synctext** [comment]
Great work, ready for a Friday email to IDEMIA!

**Comment by synctext** [comment]
PKI of IPv8 explain in easy storyline by Urbit:
This city is a social network, a peer-to-peer packet network, and a public-key infrastructure (PKI). In the social layer, your ship is your name. In the packet layer, your ship is your network address. In the PKI, your ship is your cryptographic identity. Perhaps more intuitively, a ship is a sort of pronounceable phone number.
And once Urbit is built, no one owns the city. No one can control it. It does need governance; but it governs itself as a digital republic.
Real estate transfers are low frequency and high friction, so the "double spend" problem (which mining solves) is much less pressing. https://urbit.org/posts/address-space/

**Comment by synctext** [comment]
Urbit magic: limit to 4 billion UDs. 
Most forms of network abuse are "Sybil attacks": they rely on an infinite supply of fresh identities. Scarcity makes reputation work. Spam is a business; if the cost of a new planet exceeds the amount of money you can make by spamming from that planet until its reputation is trashed, there will be no spam.
Shady stars and galaxies that sell blocks of planets to spammers will also develop reputations as "bad neighborhoods," damaging the value of the whole block. Abuse at any level is designed to be counterproductive and economically self-terminating.

**Comment by qstokkink** [comment]
I implemented a visual debugger for Twisted calls and Threads, it's fun to see what happens on the inside of IPv8:
![afbeelding](https://user-images.githubusercontent.com/3630389/35508704-a5c34d46-04f1-11e8-91af-a629be5067b5.png)

**Comment by qstokkink** [comment]
More competition in the claim domain: https://eprint.iacr.org/2018/209.pdf

**Comment by qstokkink** [comment]
I think this issue has been resolved, IPv8 is up and running. With #3615 Dispersy will be completely gone.
